src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,105 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
-/**
- * Make this class immutable. See requirements in the README file
- */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car implements Cloneable {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.engine = engine == null ? null : engine.clone();
+        this.year = year;
+        this.color = color;
+        this.wheels = cloneWheels(wheels);
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return cloneWheels(wheels);
+    }
+
+    public Engine getEngine() {
+        if (engine == null) {
+            return null;
+        }
+
+        return engine.clone();
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    public Car addWheel(Wheel wheel) {
+        List<Wheel> wheels = cloneWheels(this.wheels);
+        wheels.add(wheel);
+
+        return new Car(year, color, wheels, engine);
+    }
+
+    public Car changeColor(String color) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    private List<Wheel> cloneWheels(List<Wheel> wheels) {
+        List<Wheel> copyWheels = new ArrayList<>();
+
+        for (Wheel wheel : wheels) {
+            copyWheels.add(wheel.clone());
+        }
+
+        return copyWheels;
+    }
+
+    @Override
+    public Car clone() {
+        try {
+            Car clonedCar = (Car) super.clone();
+            clonedCar.cloneWheels(wheels);
```

Comment:
this method returns wheels clone, but you don't set it anywhere
```suggestion
            clonedCar.wheels = cloneWheels(wheels);
```